File: src/main/scala/com/vc/medium/L287.scala
package com.vc.medium

object L287 {
  def findDuplicate(arr: Array[Int]): Int = {
    var hare = arr(arr(0))
    var tortoise = arr(0)
    while(hare != tortoise) {
      hare = arr(arr(hare))
      tortoise = arr(tortoise)
    }

    tortoise = 0
    while(hare != tortoise) {
      tortoise = arr(tortoise)
      hare = arr(hare)
    }
    tortoise
  }
}

File: src/main/java/com/vc/medium/L287.java
package com.vc.medium;

class L287 {
    public int findDuplicate(int[] nums) {
        for(int i = 0; i < nums.length; i++) {
            int index = Math.abs(nums[i]) - 1;
            if(nums[index] < 0) return index + 1;
            else nums[index] = -1 * nums[index];
        }
        return -1;
    }
}
